pkg/controller/certificaterequests/venafi/venafi.go

```diff
@@ -104,31 +112,51 @@ func (v *Venafi) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerO
 		}
 	}
 
-	certPem, err := client.Sign(cr.Spec.CSRPEM, duration, customFields)
+	pickupID, exists := cr.ObjectMeta.Annotations[VenafiPickupIDAnnotation]
 
-	// Check some known error types
-	if err != nil {
-		switch err.(type) {
+	// check if the pickup ID annotation is there, if not set it up.
+	if !exists || pickupID == "" {
+		pickupID, err = client.RequestCertificate(cr.Spec.CSRPEM, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
 
-		case venafiinternal.ErrCustomFieldsType:
-			v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
-			log.Error(err, err.Error())
+			case venaficlient.ErrCustomFieldsType:
+				v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
+				log.Error(err, err.Error())
 
-			return nil, nil
+				return nil, nil
 
-		case endpoint.ErrCertificatePending:
-			message := "Venafi certificate still in a pending state, the request will be retried"
+			default:
+				message := "Failed to request venafi certificate"
 
-			v.reporter.Pending(cr, err, "IssuancePending", message)
-			log.Error(err, message)
+				v.reporter.Failed(cr, err, "RequestError", message)
+				log.Error(err, message)
+
+				return nil, err
+			}
+		}
+
+		v.reporter.Pending(cr, err, "IssuancePending", "Venafi certificate is requested")
+
+		metav1.SetMetaDataAnnotation(&cr.ObjectMeta, VenafiPickupIDAnnotation, pickupID)
+		_, err = v.cmClient.CertmanagerV1alpha2().CertificateRequests(cr.GetNamespace()).Update(ctx, cr, metav1.UpdateOptions{})
+		if err != nil {
 			return nil, err
+		}
+
+		return nil, nil
```

Comment:
I really really dislike the `Sign` function semantics when it comes to async issuers because it results in us doing things like this (`return nil, nil`) which seems just wrong 😬 .

Something for a future refactor anyway, but I'd much prefer to see this method removed as something exported and instead just integrated into the `pkg/controller/certificaterequests/venafi` package and written in a way that doesn't attempt to 'pretend' that an async process is actually synchronous (i.e. having a function named `checkIfCertificateRequestExists`, `fetchCertificateFromTPP`, all weaved together in the controller's `Reconcile` method).

Async vs sync issuers have also been a pain point that makes code confusing to reason about and read, and I think the sooner we accept that async issuers _are_ dependent on the 'controller pattern', the less indirection we'll keep imposing on ourselves.

Let's not change anything here though as this provides value beyond what we had before, and is a long-standing issue anyway :)

Reply:
I agree, with making other issuers like Vault async I think it is time to design Request() IsReady() and Fetch() kind of approach, like it is done in the actual code that talks to vCert. 

https://github.com/jetstack/cert-manager/issues/3074